From 47922b9704d0dc129cad33a8423343b16e650774 Mon Sep 17 00:00:00 2001
From: Jun Li <r65092@freescale.com>
Date: Mon, 1 Dec 2008 15:08:46 +0800
Subject: [PATCH] ENGR00088159 i.MX51 data copy between ATA&Udisk causes USB reset.

When ATA is in use and USBOTG port acts as host connects to Udisk,
ATA data access will make the USB connecttion reset. This is a USB
power supply issue, enable PMIC power supply for USB to fix it.

Signed-off-by: Li Jun <r65092@freescale.com>
---
 arch/arm/plat-mxc/utmixc.c |   19 +++++++++++++++++++
 1 files changed, 19 insertions(+), 0 deletions(-)

diff --git a/arch/arm/plat-mxc/utmixc.c b/arch/arm/plat-mxc/utmixc.c
index 409aa75..4a46b54 100644
--- a/arch/arm/plat-mxc/utmixc.c
+++ b/arch/arm/plat-mxc/utmixc.c
@@ -25,6 +25,7 @@
 #include <asm/hardware.h>
 #include <asm/arch/arc_otg.h>
 #include <asm/mach-types.h>
+#include <asm/arch/pmic_external.h>
 
 static void usb_utmi_init(struct fsl_xcvr_ops *this)
 {
@@ -58,6 +59,24 @@ static void set_power(struct fsl_xcvr_ops *this,
 			regulator_disable(usbotg_regux);
 		}
 		regulator_put(usbotg_regux, dev);
+	} else if (machine_is_mx51_3ds()) {
+		unsigned int value;
+
+		/* VUSBIN */
+		pmic_read_reg(REG_USB1, &value, 0xffffff);
+		if (on)
+			value |= 0x1;
+		else
+			value &= ~0x1;
+		pmic_write_reg(REG_USB1, value, 0xffffff);
+
+		/* VUSBEN */
+		usbotg_regux = regulator_get(dev, "USB");
+		if (on)
+			regulator_enable(usbotg_regux);
+		else
+			regulator_disable(usbotg_regux);
+		regulator_put(usbotg_regux, dev);
 	}
 }
 
-- 
1.5.4.4

